plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,267 @@
+#include "fu-wac-struct.h"
+
+struct _FuWacModuleBluetoothId9 {
+	FuWacModule parent_instance;
+};
+
+G_DEFINE_TYPE(FuWacModuleBluetoothId9, fu_wac_module_bluetooth_id9, FU_TYPE_WAC_MODULE)
+
+#define FU_WAC_MODULE_BLUETOOTH_ID9_PAYLOAD_SZ	   256
+#define FU_WAC_MODULE_BLUETOOTH_ID9_START_NORMAL   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL	   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE  0xFE
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_RAM	   0x02
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_BEGIN   0x03
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_DATA	   0x04
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL 0xEDB88320
+
+static GBytes *
+fu_wac_module_bluetooth_id9_get_startcmd(const guint8 *data,
+					 gsize sz,
+					 gboolean full_erase,
+					 GError **error)
+{
+	g_autoptr(GByteArray) loader_cmd = fu_struct_loader_cmd_new();
+	g_autoptr(GByteArray) spi_cmd = fu_struct_spi_cmd_new();
+	g_autoptr(GByteArray) unknown_cmd = fu_struct_unknown_cmd_new();
+	guint8 command = full_erase ? FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE
+				    : FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL;
+	guint32 crc;
+
+	fu_struct_unknown_cmd_set_size(unknown_cmd, sz);
+
+	fu_struct_spi_cmd_set_size(spi_cmd, sz + FU_STRUCT_UNKNOWN_CMD_SIZE);
+	if (!fu_struct_spi_cmd_set_data(spi_cmd, unknown_cmd, error)) {
+		return NULL;
+	}
+
+	fu_struct_loader_cmd_set_command(loader_cmd, command);
+	fu_struct_loader_cmd_set_size(loader_cmd, sz + FU_STRUCT_SPI_CMD_SIZE);
+	// CRC(concat(spi_cmd, *data))
+	crc = fu_crc32_full(spi_cmd->data,
+			    FU_STRUCT_SPI_CMD_SIZE,
+			    ~0,
+			    FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	crc = fu_crc32_full(data, sz, ~crc, FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	fu_struct_loader_cmd_set_crc(loader_cmd, crc);
+	if (!fu_struct_loader_cmd_set_data(loader_cmd, spi_cmd, error)) {
+		return NULL;
+	}
+
+	if (!fu_struct_loader_cmd_validate(loader_cmd->data, loader_cmd->len, 0, NULL)) {
+		return NULL;
+	}
+	return g_bytes_new(loader_cmd->data, loader_cmd->len);
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_block(FuWacModule *self,
+					guint8 phase,
+					const guint8 *data,
+					gsize sz,
+					FuProgress *progress,
+					GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GBytes) blob_chunk = NULL;
+
+	fu_byte_array_append_uint8(buf, phase);
+	g_byte_array_append(buf, data, sz);
+
+	blob_chunk = g_bytes_new(buf->data, buf->len);
+	if (!fu_wac_module_set_feature(self,
+				       FU_WAC_MODULE_COMMAND_DATA,
+				       blob_chunk,
+				       fu_progress_get_child(progress),
+				       FU_WAC_MODULE_WRITE_TIMEOUT,
+				       error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_blocks(FuWacModule *self,
+					 guint8 phase,
+					 const guint8 *data,
+					 gsize sz,
+					 gsize block_len,
+					 FuProgress *progress,
+					 GError **error)
+{
+	gsize steps = sz / block_len;
+
+	if (steps * block_len < sz) {
+		steps++;
+	}
```

Comment:
i think fu_chunk_array_new() would be better here

Reply:
Ack